Fix CPU fallback for GpuHiveTableScanExec

mythrocks opened the pull request:
Fixes #7374.

The Hive delimited text reader does not currently support nested types. When STRUCT, ARRAY, MAP, BINARY columns are encountered, the reader is expected fall back to CPU reads.

A bug in the fallback logic causes it to fail for tables with nested types, if the output projection excludes the nested type columns. This commit fixes the fallback logic.

mythrocks commented:
There is one additional issue to tackle, possibly later:

It is possible that a partition has unsupported columns that the table does not. In this case, we might still need to fall back to CPU. 

This is possibly an unlikely scenario since `TextInputFormat` tables don't seem to be `v2` data-sources. As such, dropping columns from an existing table isn't supported. So this situation isn't likely to arise.

abellina commented on integration_tests/src/main/python/hive_delimited_text_test.py:
```diff
@@ -270,24 +272,38 @@ def test_hive_text_round_trip_partitioned(spark_tmp_path, data_gen, spark_tmp_ta
 def test_hive_text_fallback_for_unsupported_types(spark_tmp_path, data_gen, spark_tmp_table_factory):
-    gen = StructGen([('my_field', data_gen)], nullable=False)
+    gen = StructGen([('my_supported_int_field', int_gen),
+                     ('my_unsupported_field', data_gen), ], nullable=False)
```
nit, if `data_gen` is all about unsupported types, think we should call it `unsupported_data_gen` or something like that. Perhaps `int_gen` could then also be renamed to `supported_data_gen`, and it can default to `int_gen`.

mythrocks replied:
Agreed. That does read better.

abellina commented on integration_tests/src/main/python/hive_delimited_text_test.py:
```diff
@@ -270,24 +272,38 @@ def test_hive_text_round_trip_partitioned(spark_tmp_path, data_gen, spark_tmp_ta
 
+    # Even if the output-projection uses only supported types,
+    # the read should fall back to CPU if the table has unsupported types.
+    assert_gpu_fallback_collect(
```
this is mostly for my own edification: we have to fall back in this scenario because the whole input is loaded to the gpu, so it is an all or nothing right? E.g. the GPU won't see an unsupported type and "skip it".

mythrocks replied:
That is correct. The whole file is consumed on GPU (i.e. all columns). The pruning of the output projection comes after the load, so we cannot do a partial read on GPU.

mythrocks replied:
P.S. That should have been clearer in the documentation. I'll add a note.

abellina commented on integration_tests/src/main/python/hive_delimited_text_test.py:
```diff
@@ -301,6 +317,7 @@ def test_hive_text_default_enabled(spark_tmp_path, data_gen, spark_tmp_table_fac
         lambda spark: read_hive_text_table(spark, table_name),
         conf={})
 
+
```
nit extra line

mythrocks replied:
Would it be alright if we kept 2 lines between functions? Without it, I see a Flake8 warning:
https://www.flake8rules.com/rules/E302.html

abellina replied:
ah sorry.. I keep forgetting this is python. Never mind

abellina reviewed (commented):
This looks good to me. I had some minor questions and a nit.

abellina commented on sql-plugin/src/main/scala/org/apache/spark/sql/hive/rapids/HiveProviderImpl.scala:
```diff
@@ -226,6 +243,10 @@ class HiveProviderImpl extends HiveProvider {
             }
 
+            // Check if datatypes are all supported.
+            tableRelation.dataCols.filter(hasUnsupportedType)
```
nit, I think we can remove the `filter` and instead invoke the `hasUnsupportedType` function within `flagUnsupportedType`, and conditionally invoke `willNotWorkOnGpu`

mythrocks replied:
Gotcha. 
I thought having them separate would be more descriptive. I've changed it now.

mythrocks commented:
Thanks for reviewing, @abellina. I've addressed the issues you raised. Could I bother you to take another look?

I've taken @revans2 off the hook for the review.

revans2 commented:
@mythrocks are we planning on waiting for nested type read support to be able to read these? or do we want to have a follow on issue that would let us only worry about the columns that we are reading?

mythrocks commented:
Sorry I missed that question, @revans2.

My plan was to leave files with nested columns as unsupported. But now that you mention it, it might be possible to read nested columns as string input, and ignore them entirely if the output projection does not include them.

revans2 commented:
I am fine either way. I would go with just falling back to the CPU at least until a customer complains.
